x/liquidstakeibc/keeper/msg_server.go

```diff
@@ -19,14 +35,294 @@ func NewMsgServerImpl(keeper Keeper) types.MsgServer {
+			}
+		default:
+			return nil, fmt.Errorf("invalid or unexpected update key: %s", update.Key)
+		}
+	}
+
+	k.SetHostChain(ctx, hc)
+
+	return &types.MsgUpdateHostChainResponse{}, nil
+}
+
+// LiquidStake defines a method for liquid staking tokens
+func (k msgServer) LiquidStake(
+	goCtx context.Context,
+	msg *types.MsgLiquidStake,
+) (*types.MsgLiquidStakeResponse, error) {
+	ctx := sdktypes.UnwrapSDKContext(goCtx)
+
+	// retrieve the host chain
+	hostChain, found := k.GetHostChainFromIbcDenom(ctx, msg.Amount.Denom)
+	if !found {
+		return nil, errorsmod.Wrapf(
+			types.ErrInvalidHostChain,
+			"host chain with ibc denom %s not registered",
+			msg.Amount.Denom,
+		)
+	}
+
+	// check for minimum deposit amount
+	if msg.Amount.Amount.LT(hostChain.MinimumDeposit) {
+		return nil, errorsmod.Wrapf(
+			types.ErrMinDeposit,
+			"expected amount more than %s, got %s",
+			hostChain.MinimumDeposit,
+			msg.Amount.Amount,
+		)
+	}
+
+	// get the delegator address from the bech32 string
+	delegatorAddress, err := sdktypes.AccAddressFromBech32(msg.DelegatorAddress)
+	if err != nil {
+		return nil, errorsmod.Wrapf(sdkerrors.ErrInvalidAddress, "error parsing delegator address: %s", err)
+	}
+
+	// amount of stk tokens to be minted
+	mintDenom := hostChain.MintDenom()
+	mintAmount := sdktypes.NewDecCoinFromCoin(msg.Amount).Amount.Mul(hostChain.CValue)
+	mintToken, _ := sdktypes.NewDecCoinFromDec(mintDenom, mintAmount).TruncateDecimal()
+
+	// send the deposit to the deposit-module account
+	depositAmount := sdktypes.NewCoins(msg.Amount)
+	err = k.bankKeeper.SendCoinsFromAccountToModule(ctx, delegatorAddress, types.DepositModuleAccount, depositAmount)
+	if err != nil {
+		return nil, errorsmod.Wrapf(
+			types.ErrFailedDeposit,
+			"failed to deposit tokens to module account %s: %s",
+			types.DepositModuleAccount,
+			err,
+		)
+	}
+
+	// update the host chain c value
+	hostChain.CValue = k.GetHostChainCValue(ctx, hostChain)
+	k.SetHostChain(ctx, hostChain)
+
+	// add the deposit amount to the deposit record for that chain/epoch
+	currentEpoch := k.GetEpochNumber(ctx, types.DelegationEpoch)
+	deposit, found := k.GetDepositForChainAndEpoch(ctx, hostChain.ChainId, currentEpoch)
+	if !found {
+		return nil, errorsmod.Wrapf(
+			types.ErrDepositNotFound,
+			"deposit not found for chain %s and epoch %v",
+			hostChain.ChainId,
+			currentEpoch,
+		)
+	}
+	deposit.Amount.Amount = deposit.Amount.Amount.Add(msg.Amount.Amount)
+	k.SetDeposit(ctx, deposit)
+
+	// mint stk tokens in the module account
+	err = k.bankKeeper.MintCoins(ctx, types.ModuleName, sdktypes.NewCoins(mintToken))
+	if err != nil {
+		return nil, errorsmod.Wrapf(
+			types.ErrMintFailed,
+			"failed to mint coins in module %s: %s",
+			types.ModuleName, err,
+		)
+	}
+
+	// calculate protocol fee
+	protocolFeeAmount := hostChain.Params.DepositFee.MulInt(mintToken.Amount)
+	protocolFee, _ := sdktypes.NewDecCoinFromDec(mintDenom, protocolFeeAmount).TruncateDecimal()
```

Comment:
this is fine.. 
But in later stages we will change the fee to collect fees in atom rather than stkatom